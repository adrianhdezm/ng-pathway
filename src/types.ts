export interface FolderMetadata {
  path: string;
  parent: string | null;
  files: string[];
}

export interface FolderNode {
  parent: string | null;
  children: FolderNode[];
  data: {
    path: string;
    files: string[];
    [key: string]: string | string[] | undefined;
  };
}

export interface Route {
  component: string;
  file?: string;
  route: string;
  children: Route[];
}

export interface RouteFile {
  fileName: string;
  fileContent: string;
}
