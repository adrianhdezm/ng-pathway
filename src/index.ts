import glob from 'glob';
import fs from 'fs';
import path from 'path';
import Handlebars from 'handlebars';
import { Route, RouteFile } from './types';
import { extractFolders, getBaseUrl } from './utils/path.utils';
import { buildFolderTreeFromHierarchy, flattenRoutes, generateFolderHierarchy, mapNodes, mapNodesToRoutes } from './utils/graph.utils';
import { computeAngularRoute, computeComponentName, isString } from './utils/string.utils';

// Construct the path to the Handlebars template file
const templatePath = path.join(__dirname, 'templates');

export function routesBuilder(pagesPattern: string): RouteFile[] {
  const filePaths = glob.sync(pagesPattern);
  const baseUrl = getBaseUrl(filePaths);

  // Extract Folders from Files Paths
  const folderPaths = extractFolders(filePaths);

  // Map Folders to Nodes with attributes Path, Parent, and Files
  const folderNodes = generateFolderHierarchy(folderPaths, baseUrl, filePaths);

  // Apply Filter to remove Folders without Files
  const filteredFolderNodes = folderNodes.filter((folderNode) => folderNode.files.length > 0);

  // Build Folder Tree From Folder Hierarchy - Create the Routes Graph
  const routeGraph = buildFolderTreeFromHierarchy(filteredFolderNodes);

  //Add Angular Route property to Nodes in Routes Graph
  const routeGraphWithAngularRoute = mapNodes(routeGraph, (node) => {
    const { data } = node;
    return {
      ...node,
      data: {
        ...data,
        route: computeAngularRoute(data.path, node.parent)
      }
    };
  });

  //Remove files from Nodes with children in Routes Graph
  const routeGraphWithFilteredFiles = mapNodes(routeGraphWithAngularRoute, (node) => {
    if (node.children.length < 1) {
      return node;
    }
    const { data } = node;
    return { ...node, data: { ...data, files: [] } };
  });

  //Add correct component file property to Nodes in Routes Graph
  const routeGraphWithFile = mapNodes(routeGraphWithFilteredFiles, (node) => {
    const { data } = node;
    return {
      ...node,
      data: {
        ...data,
        file: data.files[0]
      }
    };
  });

  // Handle Layout Nodes from Routes Graph
  const routeGraphWithoutLayoutNodes = mapNodes(routeGraphWithFile, (node) => {
    const { data, children } = node;
    const layoutNodeIndex = children.findIndex((child) => child.data.path.match(/\(([^)]+)\)/));

    if (layoutNodeIndex < 0) {
      return node;
    }

    const file = children[layoutNodeIndex].data.file;
    return {
      ...node,
      data: { ...data, file },
      children: [...children.slice(0, layoutNodeIndex), ...children.slice(layoutNodeIndex + 1)]
    };
  });

  // Compute Component Name from File Path
  const routeGraphWithComponentName = mapNodes(routeGraphWithoutLayoutNodes, (node) => {
    const { data } = node;
    return {
      ...node,
      data: {
        ...data,
        component: isString(data.file) ? computeComponentName(data.file as string) : undefined
      }
    };
  });

  // Keep only angular route relevant properties
  const routeGraphForAngularRouter = mapNodesToRoutes(routeGraphWithComponentName);

  // Map to route files

  // Read the Handlebars template from a file
  const childRoutesPartial = fs.readFileSync(path.join(templatePath, 'child-routes.hbs'), 'utf8');
  const indexTemplateSource = fs.readFileSync(path.join(templatePath, 'index.route.hbs'), 'utf8');

  // Compile the Handlebars template
  Handlebars.registerPartial('childRoutes', childRoutesPartial);
  const indexTemplate = Handlebars.compile<{
    path: string;
    component: string;
    children: Route[];
    fileImports: { file: string; component: string }[];
  }>(indexTemplateSource);

  const routeFiles = routeGraphForAngularRouter.map((node) => {
    const fileName = node.route === '' ? 'index.ts' : `${node.route}.index.ts`;

    const fileImports = [
      ...(node.file && node.component ? [{ file: node.file, component: node.component }] : []),
      ...flattenRoutes(node.children)
        .filter(({ file, component }) => file !== undefined && component !== undefined)
        .map(({ file, component }) => ({ file, component }))
    ] as { file: string; component: string }[];

    const fileContent = indexTemplate({ path: node.route, component: node.component, children: node.children, fileImports });
    return {
      path: node.route,
      fileName,
      fileContent
    };
  });

  return routeFiles;
}
