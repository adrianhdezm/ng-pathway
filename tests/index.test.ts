import glob from 'glob';
import { routesBuilder } from '../src/index';

describe('index.ts', () => {
  describe('routesBuilder', () => {
    it('renders routes for nested structures without providers', () => {
      const pagesPattern = 'src/pages/**/[a-z]*.ts';
      const files = [
        'src/pages/dashboard-page.component.ts',
        'src/pages/Teams/team-catalog-page.component.ts',
        'src/pages/Teams/[id]/team-overview-page.component.ts',
        'src/pages/Teams/[id]/(team-details)/team-details-layout.component.ts',
        'src/pages/Teams/[id]/history/team-history-page.component.ts',
        'src/pages/Teams/[id]/[...custom]/team-custom-page.component.ts',
        'src/pages/Products/product-catalog-page.component.ts',
        'src/pages/Products/[id]/product-details-page.component.ts'
      ];
      jest.spyOn(glob, 'sync').mockReturnValueOnce(files);

      const routes = routesBuilder(pagesPattern);
      expect(routes).toMatchSnapshot();
    });
  });

  it('renders routes for nested structures with providers', () => {
    const pagesPattern = 'src/pages/**/[a-z]*.ts';
    const files = [
      'src/pages/dashboard.providers.ts',
      'src/pages/dashboard-page.component.ts',
      'src/pages/Teams/team-catalog.providers.ts',
      'src/pages/Teams/team-catalog-page.component.ts',
      'src/pages/Teams/[id]/team-overview-page.component.ts',
      'src/pages/Teams/[id]/history/team-history-page.component.ts',
      'src/pages/Teams/[id]/[...custom]/team-custom-page.component.ts',
      'src/pages/Teams/[id]/(team-details)/team-details-layout.component.ts',
      'src/pages/Teams/[id]/(team-details)/team-details.providers.ts',
      'src/pages/Products/product-catalog-page.component.ts',
      'src/pages/Products/[id]/product-details-page.component.ts'
    ];
    jest.spyOn(glob, 'sync').mockReturnValueOnce(files);

    const routes = routesBuilder(pagesPattern);
    expect(routes).toMatchSnapshot();
  });

  it('renders routes for nested structures without root route', () => {
    const pagesPattern = 'src/pages/**/[a-z]*.ts';
    const files = [
      'src/pages/Teams/team-catalog.providers.ts',
      'src/pages/Teams/team-catalog-page.component.ts',
      'src/pages/Teams/[id]/team-overview-page.component.ts',
      'src/pages/Teams/[id]/team-details.providers.ts',
      'src/pages/Teams/[id]/history/team-history-page.component.ts',
      'src/pages/Teams/[id]/[...custom]/team-custom-page.component.ts',
      'src/pages/Teams/[id]/(team-details)/team-details-layout.component.ts',
      'src/pages/Products/product-catalog-page.component.ts',
      'src/pages/Products/[id]/product-details-page.component.ts'
    ];
    jest.spyOn(glob, 'sync').mockReturnValueOnce(files);

    const routes = routesBuilder(pagesPattern);
    expect(routes).toMatchSnapshot();
  });
});
