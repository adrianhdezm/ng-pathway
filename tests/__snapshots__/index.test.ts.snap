// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index.ts routesBuilder renders routes for nested structures 1`] = `
[
  {
    "fileContent": "import { Route } from '@angular/router';
import { DashboardPageComponent } from 'src/pages/dashboard-page.component';

export const ROUTES: Route[] = [
    {
        path: '',
        component: DashboardPageComponent
    }
];",
    "fileName": "index.ts",
    "path": "",
  },
  {
    "fileContent": "import { Route } from '@angular/router';
import { TeamCatalogPageComponent } from 'src/pages/Teams/team-catalog-page.component';
import { TeamDetailsLayoutComponent } from 'src/pages/Teams/[id]/(team-details)/team-details-layout.component';
import { TeamOverviewPageComponent } from 'src/pages/Teams/[id]/team-overview-page.component';
import { TeamHistoryPageComponent } from 'src/pages/Teams/[id]/history/team-history-page.component';
import { TeamCustomPageComponent } from 'src/pages/Teams/[id]/[...custom]/team-custom-page.component';

export const ROUTES: Route[] = [
    {
        path: 'Teams',
        children: [
            {
                path: '',
                component: TeamCatalogPageComponent
            },
            {
                path: ':id',
                component: TeamDetailsLayoutComponent,
                children: [
                    {
                        path: '',
                        component: TeamOverviewPageComponent
                    },
                    {
                        path: 'history',
                        component: TeamHistoryPageComponent
                    },
                    {
                        path: '*',
                        component: TeamCustomPageComponent
                    },
                ]
            },
        ]
    }
];",
    "fileName": "Teams.index.ts",
    "path": "Teams",
  },
  {
    "fileContent": "import { Route } from '@angular/router';
import { ProductCatalogPageComponent } from 'src/pages/Products/product-catalog-page.component';
import { ProductDetailsPageComponent } from 'src/pages/Products/[id]/product-details-page.component';

export const ROUTES: Route[] = [
    {
        path: 'Products',
        children: [
            {
                path: '',
                component: ProductCatalogPageComponent
            },
            {
                path: ':id',
                component: ProductDetailsPageComponent
            },
        ]
    }
];",
    "fileName": "Products.index.ts",
    "path": "Products",
  },
]
`;
